keyboards/handwired/ferris/matrix.c

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2013 Oleg Kostyuk <cub.uanic@gmail.com>
+          2020 Pierre Chevalier <pierrechevalier83@gmail.com>
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+ * This code was heavily inspired by the ergodox_ez keymap, and modernized
+ * to take advantage of the quantum.h microcontroller agnostics gpio control
+ * abstractions and use the macros defined in config.h for the wiring as opposed
+ * to repeating that information all over the place.
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <avr/io.h>
+#include "wait.h"
+#include "action_layer.h"
+#include "print.h"
+#include "debug.h"
+#include "util.h"
+#include "matrix.h"
+#include "debounce.h"
+#include QMK_KEYBOARD_H
+
+/* matrix state(1:on, 0:off) */
+static matrix_row_t raw_matrix[MATRIX_ROWS];  // raw values
+static matrix_row_t matrix[MATRIX_ROWS];      // debounced values
+
+static matrix_row_t read_cols(uint8_t row);
+static void         init_cols(void);
+static void         unselect_rows(void);
+static void         select_row(uint8_t row);
+
+static uint8_t mcp23017_reset_loop;
+
+void matrix_init_custom(void) {
+    // initialize row and col
+
+    mcp23017_status = init_mcp23017();
+
+    unselect_rows();
+    init_cols();
+
+    // initialize matrix state: all keys off
+    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+        matrix[i]     = 0;
+        raw_matrix[i] = 0;
+    }
+}
+
+void matrix_power_up(void) {
+    mcp23017_status = init_mcp23017();
+
+    unselect_rows();
+    init_cols();
+
+    // initialize matrix state: all keys off
+    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+        matrix[i] = 0;
+    }
+}
+
+// Reads and stores a row, returning
+// whether a change occurred.
+static inline bool store_matrix_row(matrix_row_t current_matrix[], uint8_t index) {
+    matrix_row_t temp = read_cols(index);
+    if (current_matrix[index] != temp) {
+        current_matrix[index] = temp;
+        return true;
+    }
+    return false;
+}
+
+bool matrix_scan_custom(matrix_row_t current_matrix[]) {
+    if (mcp23017_status) {  // if there was an error
+        if (++mcp23017_reset_loop == 0) {
+            // if (++mcp23017_reset_loop >= 1300) {
+            // since mcp23017_reset_loop is 8 bit - we'll try to reset once in 255 matrix scans
+            // this will be approx bit more frequent than once per second
+            print("trying to reset mcp23017\n");
```

Comment:
Make all these logging statements into `dprint()`s, and you can get rid of the print.h include.